main/src/ca/uwaterloo/flix/language/errors/NonExhaustiveMatchError.scala

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2017 Jason Mitterteiner
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.errors
+
+import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast.{SourceInput, SourceLocation, TypedAst}
+import ca.uwaterloo.flix.util.vt.VirtualString._
+import ca.uwaterloo.flix.util.vt.VirtualTerminal
+
+/**
+  * An error raised to indicate a non exhaustive pattern match
+  *
+  */
+case class NonExhaustiveMatchError(rules: List[TypedAst.MatchRule], pattern: String, loc: SourceLocation) extends CompilationError {
+  val kind = "Exhaustive Match Error"
+  val source: SourceInput = loc.source
+  val message: VirtualTerminal = {
+    val vt = new VirtualTerminal
+    vt << Line(kind, source.format) << NewLine
+    vt << ">> Exhaustive Match Error:" << NewLine
+    vt << NewLine
+    vt << "The expression: " << NewLine
+    vt << Code(loc, "")
+    vt << "Matched by the rules:" << NewLine
```

Comment:
It doesn't print out the full match, just the pattern. I think it makes sense to do that since it allows you to look at the cases and visually confirm for yourself the it is missing something. However, it can get verbose and I can see it either way.

The full error message looks like:

```
>> Exhaustive Match Error:

The expression: 
218 | def evalInst(instructions: List[Inst], stack: List[Int]): Int = match (instructions, stack) with {
                                                                            ^^^^^^^^^^^^^^^^^^^^^

Matched by the rules:
219 |     case (Nil, x :: _) => x
               ^^^^^^^^^^^^^
               
220 |     case ((Push(i)) :: rs, st) => evalInst(rs, i :: st)
               ^^^^^^^^^^^^^^^^^^^^^
               
221 |     case (Add :: rs, i1 :: i2 :: st) => evalInst(rs, (i1 + i2) :: st)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
               
222 |     case (Sub :: rs, i1 :: i2 :: st) => evalInst(rs, (i1 - i2) :: st)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
               
223 |     case (Mul :: rs, i1 :: i2 :: st) => evalInst(rs, (i1 * i2) :: st)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
               
224 |     case (Div :: rs, i1 :: i2 :: st) => evalInst(rs, (i1 / i2) :: st)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
               
225 |     case (Neg :: rs, i :: st) =>
               ^^^^^^^^^^^^^^^^^^^^
               
230 |     case (And :: rs, i1 :: i2 :: st) =>
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
               
235 |     case (Or :: rs, i1 :: i2 :: st) =>
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
               
240 |     case (Cmp :: rs, i1 :: i2 :: st) =>
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^
               
245 |     case ((Branch(is1, is2)) :: rs, i :: st) =>
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
               
is not exhaustive, consider the pattern: 
(Nil, Nil)
```